### PR TITLE
Deterministic 2

### PR DESCRIPTION
**Pull request by larsgottesbueren**
This PR contains

1) some non-determinism fixes in the coarsener
2) a parallelization of the block-pair-wise move approval algorithm
3) adds fixed vertex support for the deterministic code
4) adds the new gain types to the deterministic code

+ removes some old unused code.

### CODE REVIEW

File: mt-kahypar/datastructures/concurrent_bucket_map.h [hunk 1]
```diff
@@ -69,8 +69,7 @@ class ConcurrentBucketMap {
  public:
 
   ConcurrentBucketMap() :
-    _num_buckets(align_to_next_power_of_two(
-      BUCKET_FACTOR * std::thread::hardware_concurrency())),
+    _num_buckets(align_to_next_power_of_two(BUCKET_FACTOR * BUCKET_FACTOR)),
```
Comment (N-Maas) [inline mt-kahypar/datastructures/concurrent_bucket_map.h:72]: Is this change actually necessary? It seems to me that with the adjustment in `static_hypergraph` all code that uses this is already deterministic. I think it makes sense to scale the number of buckets with the number of threads

**Review by N-Maas — CHANGES_REQUESTED** [review]
Looks good overall.

The test coverage for `deterministic_label_propagation` is rather low, perhaps you could extend `determinism_test` a bit? (e.g., a test where `sync-lp-recalculate-gains-on-second-apply` is true)

**Comment by larsgottesbueren** [comment]
> Looks good overall.

Thanks

> The test coverage for `deterministic_label_propagation` is rather low, perhaps you could extend `determinism_test` a bit? (e.g., a test where `sync-lp-recalculate-gains-on-second-apply` is true)

I will remove this particular option since it didn't give any benefit back then.

File: mt-kahypar/datastructures/concurrent_bucket_map.h (hunk 1 above)
Comment (larsgottesbueren) [reply mt-kahypar/datastructures/concurrent_bucket_map.h:72]: I think this is necessary to get deterministic results across different machines. What do you think about letting the caller determine the value?

File: mt-kahypar/io/command_line_options.cpp [hunk 2]
```diff
@@ -72,8 +72,8 @@ namespace mt_kahypar {
     options.add_options()
             ("help", "show help message")
             ("deterministic", po::value<bool>(&context.partition.deterministic)->value_name("<bool>")->default_value(false),
-             "Shortcut to enables deterministic partitioning mode, where results are reproducible across runs. "
-             "If set, the specific deterministic subroutines don't need to be set manually.")
+             "Enables determinism in preprocessing in initial partitioning. "
```
[the remark this answers is not in the document]
Comment (larsgottesbueren) [reply mt-kahypar/io/command_line_options.cpp]: Fixed. The current behavior is now enabling determinism overall again.

File: mt-kahypar/datastructures/static_hypergraph.h [hunk 3]
```diff
@@ -799,6 +799,7 @@ class StaticHypergraph {
     ASSERT(edgeIsEnabled(he), "Hyperedge" << he << "is disabled");
     const size_t incidence_array_start = hyperedge(he).firstEntry();
     const size_t incidence_array_end = hyperedge(he).firstInvalidEntry();
+    // TODO this breaks if we have multi-pins. same for restore. we don't have a check for multi-pins in the input
```
[the remark this answers is not in the document]
Comment (larsgottesbueren) [reply mt-kahypar/datastructures/static_hypergraph.h]: This was an old comment (from 2021) when we didn't have this check. Thanks for catching this.

File: mt-kahypar/datastructures/static_hypergraph.cpp [hunk 4]
```diff
@@ -288,6 +286,10 @@ namespace mt_kahypar::ds {
           // Update number of incident nets of high degree vertex
           const size_t contracted_size = incident_nets_pos.load() - incident_nets_start;
           tmp_hypernodes[coarse_hn].setSize(contracted_size);
+
+          // sort for determinism
+          tbb::parallel_sort(tmp_incident_nets.begin() + incident_nets_start,
+                             tmp_incident_nets.begin() + incident_nets_start + contracted_size);
```
[the remark this answers is not in the document]
Comment (larsgottesbueren) [reply mt-kahypar/datastructures/static_hypergraph.cpp]: I added a flag `deterministic` so we only do it then. There is also an old branch with an improved contraction implementation, that never got finished. If we're concerned about performance, we could look at that again.

File: mt-kahypar/datastructures/concurrent_bucket_map.h (hunk 1 above)
Comment (N-Maas) [reply mt-kahypar/datastructures/concurrent_bucket_map.h:72]: Seems like a good idea. With regards to determinism, I thought that the algorithms that currently use the `ConcurrentBucketMap` are such that the output does not depend on the distribution to the buckets. If the caller determines the value, we can decide this for each algorithm separately :+1:

File: mt-kahypar/partition/refinement/deterministic/deterministic_label_propagation.h [hunk 5]
```diff
@@ -71,6 +72,7 @@ class DeterministicLabelPropagationRefiner final : public IRefiner {
 
 private:
   static constexpr bool debug = false;
+  static constexpr size_t invalid_pos = std::numeric_limits<size_t>::max() / 2;
```
[the remark this answers is not in the document]
Comment (larsgottesbueren) [reply mt-kahypar/partition/refinement/deterministic/deterministic_label_propagation.h:72]: I don't remember tbh. Usually this means that I wanted to add two boundaries. At the moment it doesn't look like that's the case any more.

File: mt-kahypar/datastructures/concurrent_bucket_map.h (hunk 1 above)
Comment (larsgottesbueren) [reply mt-kahypar/datastructures/concurrent_bucket_map.h:72]: I looked at this again and you were right. The two uses in StaticHypergraph::contract(..) don't depend on the distribution of elements to buckets. So I reverted the change :)

File: mt-kahypar/datastructures/static_hypergraph.cpp [hunk 6]
```diff
@@ -394,7 +397,7 @@ namespace mt_kahypar::ds {
       // Compute start position of each hyperedge in incidence array
       parallel::TBBPrefixSum<size_t, Array> num_pins_prefix_sum(he_sizes);
       tbb::parallel_invoke([&] {
-        tbb::parallel_for(ID(0), _num_hyperedges, [&](const HyperedgeID& id) {
+        tbb::parallel_for(size_t(0), size_t(_num_hyperedges), [&](const HyperedgeID& id) {
```
Comment (kittobi1992) [inline mt-kahypar/datastructures/static_hypergraph.cpp:400]: HyperedgeID is uint32_t, why we use here size_t?

File: mt-kahypar/datastructures/static_hypergraph.cpp [hunk 7]
```diff
@@ -414,11 +417,11 @@ namespace mt_kahypar::ds {
       // Write hyperedges from temporary buffers to incidence array
       tbb::enumerable_thread_specific<size_t> local_max_edge_size(UL(0));
       tbb::parallel_for(ID(0), _num_hyperedges, [&](const HyperedgeID& id) {
-        if ( he_mapping.value(id) /* hyperedge is valid */ ) {
+        if ( he_mapping.value(id) > 0 /* hyperedge is valid */ ) {
           const size_t he_pos = he_mapping[id];
           const size_t incidence_array_start = num_pins_prefix_sum[id];
           Hyperedge& he = hypergraph._hyperedges[he_pos];
-          he = std::move(tmp_hyperedges[id]);
+          he = tmp_hyperedges[id];
```
Comment (kittobi1992) [inline mt-kahypar/datastructures/static_hypergraph.cpp:424]: Why removing std::move(...) here?

File: mt-kahypar/partition/coarsening/deterministic_multilevel_coarsener.cpp [hunk 8]
```diff
@@ -97,7 +99,7 @@ bool DeterministicMultilevelCoarsener<TypeTraits>::coarseningPassImpl() {
   if (num_nodes_before_pass / num_nodes <= _context.coarsening.minimum_shrink_factor) {
     return false;
   }
-  _uncoarseningData.performMultilevelContraction(std::move(clusters), pass_start_time);
+  _uncoarseningData.performMultilevelContraction(std::move(clusters), /*deterministic=*/true, pass_start_time);
```
Comment (kittobi1992) [inline mt-kahypar/partition/coarsening/deterministic_multilevel_coarsener.cpp:102]: `true /* deterministic */`

File: mt-kahypar/partition/coarsening/multilevel_coarsener.h [hunk 9]
```diff
@@ -213,7 +213,7 @@ class MultilevelCoarsener : public ICoarsener,
 
     _timer.start_timer("contraction", "Contraction");
     // Perform parallel contraction
-    _uncoarseningData.performMultilevelContraction(std::move(cluster_ids), round_start);
+    _uncoarseningData.performMultilevelContraction(std::move(cluster_ids), /*deterministic=*/false, round_start);
```
Comment (kittobi1992) [inline mt-kahypar/partition/coarsening/multilevel_coarsener.h:216]: `false /* deterministic */`

File: mt-kahypar/partition/refinement/deterministic/deterministic_label_propagation.cpp [hunk 10]
```diff
@@ -28,27 +28,30 @@
...
           Metrics& best_metrics,
           const double) {
     PartitionedHypergraph& phg = utils::cast<PartitionedHypergraph>(hypergraph);
     Gain overall_improvement = 0;
     constexpr size_t num_buckets = utils::ParallelPermutation<HypernodeID>::num_buckets;
-    const size_t num_sub_rounds = context.refinement.deterministic_refinement.num_sub_rounds_sync_lp;
-    const size_t num_buckets_per_sub_round = parallel::chunking::idiv_ceil(num_buckets, num_sub_rounds);
+    size_t num_sub_rounds = context.refinement.deterministic_refinement.num_sub_rounds_sync_lp;
+
+    using GainComputation = typename GainTypes::GainComputation;
```
Comment (kittobi1992) [inline mt-kahypar/partition/refinement/deterministic/deterministic_label_propagation.cpp:53]: define this at class scope

**Review by kittobi1992 — CHANGES_REQUESTED** [review]
LGTM, only some nitpicks. 

Currently, using fixed vertices in deterministic mode will lead to a `NonSupportedOperationException`. Remove the corresponding restriction in `partitioner.cpp`. Also make sure that using another gain function in deterministic mode will not silently switch to km1. I have not found the corresponding part in the code where this is enforced. If this is already possible ignore this comment.

File: mt-kahypar/datastructures/static_hypergraph.cpp (hunk 6 above)
Comment (larsgottesbueren) [reply mt-kahypar/datastructures/static_hypergraph.cpp]: Changed to HyperedgeID cast

File: mt-kahypar/datastructures/static_hypergraph.cpp (hunk 7 above)
Comment (larsgottesbueren) [reply mt-kahypar/datastructures/static_hypergraph.cpp:424]: Because it's slightly misleading. The type Hyperedge consists only of POD types so it's just a copy.

File: mt-kahypar/partition/coarsening/deterministic_multilevel_coarsener.cpp (hunk 8 above)
Comment (larsgottesbueren) [reply mt-kahypar/partition/coarsening/deterministic_multilevel_coarsener.cpp]: Done

File: mt-kahypar/partition/refinement/deterministic/deterministic_label_propagation.cpp (hunk 10 above)
Comment (larsgottesbueren) [reply mt-kahypar/partition/refinement/deterministic/deterministic_label_propagation.cpp]: Changed to definition at class scope. Btw, a nice effect of doing this locally is that we're not exposing the gain_definitions.h in the header. Slimmer headers lead to faster compile times.

File: mt-kahypar/partition/coarsening/multilevel_coarsener.h (hunk 9 above)
Comment (larsgottesbueren) [reply mt-kahypar/partition/coarsening/multilevel_coarsener.h]: Done

**Comment by larsgottesbueren** [comment]
> LGTM, only some nitpicks.
> 
> Currently, using fixed vertices in deterministic mode will lead to a `NonSupportedOperationException`. Remove the corresponding restriction in `partitioner.cpp`. Also make sure that using another gain function in deterministic mode will not silently switch to km1. I have not found the corresponding part in the code where this is enforced. If this is already possible ignore this comment.

There is no silent switch to km1. There is a silent switch to the deterministic refiner, but the refiner names don't pertain to an objective function (as in sequential kahypar).

File: mt-kahypar/partition/refinement/deterministic/deterministic_label_propagation.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (larsgottesbueren) [reply mt-kahypar/partition/refinement/deterministic/deterministic_label_propagation.h:75]: I also see no harm in keeping it this way btw :)

File: mt-kahypar/partition/refinement/deterministic/deterministic_label_propagation.cpp [hunk 11]
```diff
@@ -38,17 +37,19 @@
...
           Metrics& best_metrics,
           const double) {
     PartitionedHypergraph& phg = utils::cast<PartitionedHypergraph>(hypergraph);
     Gain overall_improvement = 0;
     constexpr size_t num_buckets = utils::ParallelPermutation<HypernodeID>::num_buckets;
-    const size_t num_sub_rounds = context.refinement.deterministic_refinement.num_sub_rounds_sync_lp;
-    const size_t num_buckets_per_sub_round = parallel::chunking::idiv_ceil(num_buckets, num_sub_rounds);
+    size_t num_sub_rounds = context.refinement.deterministic_refinement.num_sub_rounds_sync_lp;
+
+    GainComputation gain_computation(context, true);
```
Comment (N-Maas) [inline mt-kahypar/partition/refinement/deterministic/deterministic_label_propagation.cpp:52]: The other refinement algorithms use a member variable for `GainComputation`. I think this should preferably be uniform.

Since the steiner tree gain computation actually contains allocations that can be reused, I think making it a member makes sense and should also be done for the deterministic LP

**Review by N-Maas — APPROVED** [review]
LGTM, only the issue with `GainComputation` should be fixed before merging.
